http-clients/src/main/java/feign/ByteArrayDelegateDecoder.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feign;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.net.HttpHeaders;
+import feign.codec.DecodeException;
+import feign.codec.Decoder;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Collection;
+import javax.ws.rs.core.MediaType;
+
+/**
+ * Decodes to a byte array if the response has a Content-Type of application/octet-stream and the return type is a byte
+ * array, or falls back to the given delegate otherwise.
+ */
+public final class ByteArrayDelegateDecoder implements Decoder {
+    private final Decoder delegate;
+
+    public ByteArrayDelegateDecoder(Decoder delegate) {
+        this.delegate = delegate;
+    }
+
+
+    @Override
+    public Object decode(Response response, Type type) throws IOException, DecodeException, FeignException {
+        Collection<String> contentTypes = response.headers().get(HttpHeaders.CONTENT_TYPE);
+        if (contentTypes == null) {
+            contentTypes = ImmutableSet.of();
+        }
+        // In the case of multiple content types, or an unknown content type, we'll use the delegate instead.
+        if (contentTypes.size() == 1 && type.equals(byte[].class)
+                && Iterables.getOnlyElement(contentTypes, "contentTypes").equals(MediaType.APPLICATION_OCTET_STREAM)) {
```

Comment:
thinking this through, I think we should maybe skip this filter -- if your class expects a byte[], we should hand it the byte[]. Also wondering if we should special case handling of `InputStream`, so that if your class expects an `InputStream` we give it the body `InputStream`?